core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

```diff
@@ -69,13 +69,15 @@ import org.apache.spark.util.Utils
  * [options] represent the specific property of this source or sink.
  */
 private[spark] class MetricsSystem private (
-    val instance: String, conf: SparkConf) extends Logging {
+    val instance: String,
+    conf: SparkConf,
+    registry: MetricRegistry)
+  extends Logging {
 
   private[this] val metricsConfig = new MetricsConfig(conf)
 
   private val sinks = new mutable.ArrayBuffer[Sink]
-  private val sources = new mutable.ArrayBuffer[Source]
-  private val registry = new MetricRegistry()
+  private val sourcesWithListeners = new mutable.HashMap[Source, MetricRegistryListener]
```

Comment:
```suggestion
  private val sourcesWithListeners = new mutable.ArrayBuffer[(Source, MetricRegistryListener)]
```

Let us preserve the current semantics and not introduce assumption that Source has consistent `equals`/`hashCode`.

Reply:
+1 for the above comment.